### ISSUE TITLE
Database schema sharing (multiple modules)

### ISSUE THREAD

**Issue by leonkosak** [abpframework/abp#2511]
For instance for BookStore-Modular: https://github.com/abpframework/abp/tree/dev/samples/BookStore-Modular

How is the most efficient way to "share" part of  "whole database schema" across multiple modules (so that some tables are not defined on application-level but only on module level)?

If application would have the same architecture as BookStore-Modular and part od database would be using the same tables in 3 modules (and other parts of the application - modules don't use these tables at all).

Maybe module, which "holds" tables for modules and each module that should use these tables should reference this "database module"?

Thank you for suggestions.

**Comment by hikalkan** [comment]
>Maybe module, which "holds" tables for modules and each module that should use these tables should reference this "database module"?

It should be something like that. You need a shared module which defines the shared entities. I have an idea to create a documentation for these kind of use cases, by the time.

**Comment by leonkosak** [comment]
@hikalkan: https://docs.abp.io/en/abp/latest/Connection-Strings

Based on this document - if module has defined "[ConnectionStringName("AbpIdentityServer")]" and on application-level project, there is no definition for "AbpIdentityServer", Migrator will create tables for such module in "Default" connection string database?

**Comment by leonkosak** [comment]
![image](https://user-images.githubusercontent.com/39346964/71816757-ec358180-3083-11ea-8579-5eecfa687dab.png)
Crap, this is answer on my question. :) Sorry.

**Comment by hikalkan** [comment]
>When creating a new tenant, all tables would be inside a dedicated database (for the tenant)?

We suggest to separate tenant and host applications, and separate tenant and host dbcontextes. In this was, schemas (tables) will be different. But the default startup template doesn't implement this since it brings additional complexity, requires a deeper knowledge and should be designed by you :)

**Comment by leonkosak** [comment]
Why different SQL schemas (tables) for host and tenant databases? The schema has to be the same for all tables (just like ANZ - the hot database has the same schema as tenant databases).

**Comment by hikalkan** [comment]
>Why different SQL schemas (tables) for host and tenant databases

Because tenant side should not have "Tenants" or "Editions" table as a best practice :) You can ignore it and create empty, never used tables in every tenant database. We did it in Zero since it was simpler to implement and use. For ABP.IO, we are trying to allow to implement best practices better. However, it is completely up to you.

**Comment by leonkosak** [comment]
Yes - I have in mind empty tables for "Tenants", "Editions" and other host-related tables in tenant databases. :)
Is there any documentation how one SQL table should be in host database (and not in tenant database and vice versa)?
https://docs.abp.io/en/abp/latest/Multi-Tenancy
If class (table) inherits from IMultiTenant, the table is for tenants, but Guid (TenantId) is still nullable (which probably means "host record")?

**Comment by hikalkan** [comment]
When you have 2 dbcontext and 2 db migrations, you will have 2 different table set by nature. Just don't call `builder.ConfigureTenantManagement();` ([see](https://github.com/abpframework/abp/blob/dev/templates/app/aspnet-core/src/MyCompanyName.MyProjectName.EntityFrameworkCore.DbMigrations/EntityFrameworkCore/MyProjectNameMigrationsDbContext.cs#L41)) in the tenant dbcontext, call it in the host dbcontext.

**Comment by leonkosak** [comment]
@hikalkan I successfully created a solution structure from scratch based on _BookStore-Modular_ example.
Currently, there is one defined entity (for SQL table) in module (in .Domain project in a module) and added DbSet in module .EntityFrameworkCore project (in [ModuleName]DbContext.cs file).

It's desired, that all entities (SQL tables) from all modules would be in one SQL database (and each tenant would have it's own instance of such database with the same schema).

But what I haven't found answer is this:

1) I added builder.Configure[ModuleName] in [MyApplication]NameMigrationsDbContext.cs (.EntityFrameworkCore.DbMigrations project **on application level**)
I see configurations and settings ConnectionStringName, DbTablePrefix, DbSchema (in mudule solution) but I am missing something that table is not generated migration script.
Of course -  on application level, there is only "Default" connection string defined.
What am I doing wrong?

2) If sometime in the future I want to go microservice approach (and separated deployments for modules and therefore also separated databases for each module), then migrations should be generated on module level?
Currently, I haven't succeeded in this action because of an error:
Startup project 'src\ModuleDatabase.EntityFrameworkCore' targets framework '.NETStandard'. There is no runtime associated with this framework, and projects targeting it cannot be executed directly. To use the Entity Framework Core Package Manager Console Tools with this project, add an executable project targeting .NET Framework or .NET Core that references this project, and set it as the startup project; or, update this project to cross-target .NET Framework or .NET Core. For more information on using the EF Core Tools with .NET Standard projects, see https://go.microsoft.com/fwlink/?linkid=2034705

Thank you
